src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelInternal.java

```diff
@@ -433,43 +434,37 @@ public String getLatestCommittedOffsetToken() {
     return response.getPersistedOffsetToken();
   }
 
-  /**
-   * Check whether we need to throttle the insert API under the following low memory condition:
-   * <li>system free_memory/total_memory < INSERT_THROTTLE_THRESHOLD_IN_PERCENTAGE
-   */
+  /** Check whether we need to throttle the insertRows API */
   void throttleInsertIfNeeded(Runtime runtime) {
-    int insertThrottleThresholdInPercentage =
-        this.owningClient.getParameterProvider().getInsertThrottleThresholdInPercentage();
-    if (runtime.freeMemory() * 100 / runtime.totalMemory() < insertThrottleThresholdInPercentage) {
-      long oldTotalMem = runtime.totalMemory();
-      long oldFreeMem = runtime.freeMemory();
-      int retry = 0;
-
-      long insertThrottleIntervalInMs =
-          this.owningClient.getParameterProvider().getInsertThrottleIntervalInMs();
-      while (runtime.freeMemory() * 100 / runtime.totalMemory()
-              < insertThrottleThresholdInPercentage
-          && retry < INSERT_THROTTLE_MAX_RETRY_COUNT) {
-        try {
-          Thread.sleep(insertThrottleIntervalInMs);
-          retry++;
-        } catch (InterruptedException e) {
-          throw new SFException(ErrorCode.INTERNAL_ERROR, "Insert throttle get interrupted");
-        }
+    int retry = 0;
+    long insertThrottleIntervalInMs =
+        this.owningClient.getParameterProvider().getInsertThrottleIntervalInMs();
+    while ((hasLowRuntimeMemory(runtime)
+            || this.owningClient.getFlushService().throttleDueToQueuedFlushTasks())
+        && retry < INSERT_THROTTLE_MAX_RETRY_COUNT) {
```

Comment:
My concern is that the throttling has 2 reasons to stop: either throttled successfully or give up because of too many attempts. In the second case, it may lead to OOM errors anyways.

Reply:
What will happen then when the connectivity to Snowflake is really poor, i.e. it cannot transfer more than 10kB/s, but `insertRows` is being called with the rate of 10MB/s? Will we encounter OOM eventually?